hawc/apps/animal/managers.py

```diff
@@ -297,6 +298,105 @@ def migrate_terms(self, assessment: Assessment) -> pd.DataFrame:
+                raise ValidationError("Each endpoint must have an 'id'")
+            if len(obj.keys()) < 2:
+                raise ValidationError(f"No term fields given for endpoint {obj['id']}")
+
+        # ids must be unique
+        endpoint_ids = [obj["id"] for obj in objs]
+        if len(endpoint_ids) != len(set(endpoint_ids)):
+            raise ValidationError("Endpoint ids must be unique")
+        # ids must be valid
+        endpoints = self.get_queryset().filter(pk__in=endpoint_ids)
+        valid_endpoint_ids = endpoints.values_list("id", flat=True)
+        invalid_endpoint_ids = set(endpoint_ids) - set(valid_endpoint_ids)
+        if len(invalid_endpoint_ids) > 0:
+            invalid_endpoint_ids_str = ", ".join(str(_) for _ in invalid_endpoint_ids)
+            raise ValidationError(f"Invalid id(s) {invalid_endpoint_ids_str}")
+        # endpoints must be in the same assessment
+        assessment = endpoints.first().assessment
+        invalid_endpoints = endpoints.exclude(assessment_id=assessment.id)
+        if invalid_endpoints.exists():
+            invalid_endpoints_str = ", ".join(str(_.pk) for _ in invalid_endpoints)
+            raise ValidationError(
+                f"Endpoints must be from the same assessment; endpoint(s) {invalid_endpoints_str} not from assessment {assessment.id}"
+            )
+        # assessment must have vocab
+        if assessment.vocabulary is None:
+            raise ValidationError("Vocabulary not set in assessment {assessment.id}")
+        # restructure dict for easier seperation of id and terms
+        endpoint_id_to_terms = {obj.pop("id"): obj for obj in objs}
+        # obj fields must be terms
+        term_fields = {
+            "system_term_id",
+            "organ_term_id",
+            "effect_term_id",
+            "effect_subtype_term_id",
+            "name_term_id",
+        }
+        for term_field_to_id in endpoint_id_to_terms.values():
+            if not set(term_field_to_id.keys()).issubset(term_fields):
+                invalid_fields_str = ", ".join(set(term_field_to_id.keys()) - term_fields)
+                raise ValidationError(f"Invalid field name(s) '{invalid_fields_str}'")
+        # all terms must be in assessment vocabulary
+        term_fields_and_ids = {
+            term_field_and_id
+            for term_field_to_id in endpoint_id_to_terms.values()
+            for term_field_and_id in term_field_to_id.items()
+        }
+        term_ids = [id for (_, id) in term_fields_and_ids]
+        terms = Term.objects.filter(pk__in=term_ids)
+        for term in terms:
+            if term.namespace != assessment.vocabulary:
+                raise ValidationError(
+                    f"Assessment vocabulary {assessment.vocabulary} does not match term {term.id} namespace {term.namespace}"
+                )
+        # all terms must match assigned type
+        term_types = VocabularyTermType.as_dict()
+        term_type_to_field = {
```

Comment:
you've used these maps in lots of different methods - can we refactor this and put these maps somewhere shared?